quadtree.js

```diff
@@ -273,74 +322,31 @@ class QuadTree {
     return found;
   }
 
-  closest(point, count, maxDistance) {
+  closest(point, count = 1, maxDistance = Infinity, furthestFound = Infinity, foundSoFar = 0) {
     if (typeof point === "undefined") {
       throw TypeError("Method 'closest' needs a point");
     }
-    if (typeof count === "undefined") {
-      count = 1;
-    }
 
-    // Limit to number of points in this QuadTree
-    if (this.length == 0) {
-      return [];
-    }
-    if (this.length < count) {
-      return this.points;
-    }
+    var inRangePoints = [...this.points].filter((p) => p.distanceFrom(point) <= maxDistance);
 
-    if (typeof maxDistance === "undefined") {
-      // A circle as big as the QuadTree's boundary
-      const outerReach = Math.sqrt(
-        Math.pow(this.boundary.w / 2, 2) + Math.pow(this.boundary.h / 2, 2)
-      );
-      // Distance of query point from center
-      const pointDistance = Math.sqrt(
-        Math.pow(point.x, 2) + Math.pow(point.y, 2)
-      );
-      // Together, a circle that encompasses the whole QuadTree
-      maxDistance = outerReach + pointDistance;
-    } else {
-      // Make sure the largest search (maxDistance) contains enough points
-      let maxOuter = new Circle(point.x, point.y, maxDistance);
-      let maxDistanceTest = this.query(maxOuter);
-      if (maxDistanceTest.length < count) {
-        return maxDistanceTest;
-      }
-    }
+    let furthestDistance = 0;
 
-    // Binary search with Circle queries
-    let inner = 0;
-    let outer = maxDistance;
-    let limit = 8; // Limit to avoid infinite loops caused by ties
-    let points;
-    while (limit > 0) {
-      const radius = (inner + outer) / 2;
-      const range = new Circle(point.x, point.y, radius);
-      points = this.query(range);
-      if (points.length === count) {
-        return points; // Return when we hit the right size
-      } else if (points.length < count) {
-        // Grow
-        inner = radius;
-      } else {
-        // Shrink
-        outer = radius;
-        limit --;
-      }
-    }
-    // Sort by squared distance
-    points.sort(
-      (a, b) => {
-        const aDist = Math.pow(point.x - a.x, 2) +
-          Math.pow(point.y - a.y, 2);
-        const bDist = Math.pow(point.x - b.x, 2) +
-          Math.pow(point.y - b.y, 2);
-        return aDist - bDist;
+    inRangePoints.forEach(p => {
+        furthestDistance = Math.max(p.distanceFrom(point), furthestDistance);
+    });
+
+    furthestDistance = Math.min(furthestDistance, furthestFound);
+
+    this.children.forEach(child => {
+      if ((child.boundary.distanceFrom(point) < furthestFound || foundSoFar < count) &&
+          child.boundary.distanceFrom(point) < maxDistance) {
```

Comment:
Style nitpick, how do you like the looks of this, which better matches the code styles elsewhere.

```JS
if (
  (blah || blah) &&
  blah
) {
  // code
}